### PR TITLE
feat(joins): support MySQL 8.0

### PR DESCRIPTION
**Pull request by Weakky**
## Overview

closes https://github.com/prisma/team-orm/issues/455

Adds support for joins on MySQL 8.0, through correlated sub-queries. Also works with PlanetScale. Does not work with MySQL 5.6, MySQL 5.7 and MariaDB.

Here are examples for 1-1, 1-m and m-n relations between a `User` and `Post` model:

### 1-1

```prisma
model User {
  id   Int     @id @default(autoincrement())
  name String?

  posts Post[]
}

model Post {
  id    Int     @id @default(autoincrement())
  title String?

  user   User? @relation(fields: [userId], references: [id])
  userId Int?
}
```

```graphql
{
  findManyPost {
    id
    title
    user {
      id
      name
    }
  }
}
```

```sql
SELECT
  `t1`.`id`,
  `t1`.`title`,
  `t1`.`userId`,
  (
    SELECT
      JSON_OBJECT('id', `t2`.`id`, 'name', `t2`.`name`)
    FROM
      `User` AS `t2`
    WHERE
      `t1`.`userId` = `t2`.`id`
    LIMIT
      1
  ) AS `user`
FROM
  `Post` AS `t1`
```

### 1-m

```prisma
model User {
  id   Int     @id @default(autoincrement())
  name String?

  posts Post[]
}

model Post {
  id    Int     @id @default(autoincrement())
  title String?

  user   User? @relation(fields: [userId], references: [id])
  userId Int?
}
```

```graphql
{
  findManyUser {
    id
    name
    posts(orderBy: { id: asc }) {
      id
      title
    }
  }
}
```

```sql
SELECT
  `t1`.`id`,
  `t1`.`name`,
  (
    SELECT
      COALESCE(
        JSON_ARRAYAGG(`__prisma_data__`),
        CONVERT('[]', JSON)
      ) AS `__prisma_data__`
    FROM
      (
        SELECT
          `t4`.`__prisma_data__`
        FROM
          (
            SELECT
              JSON_OBJECT('id', `t3`.`id`, 'title', `t3`.`title`) AS `__prisma_data__`,
              `t3`.`id`
            FROM
              (
                SELECT
                  `t2`.*
                FROM
                  `Post` AS `t2`
                WHERE
                  `t1`.`id` = `t2`.`userId`
                  /* root select */
              ) AS `t3`
              /* inner select */
          ) AS `t4`
        ORDER BY
          `t4`.`id` ASC
        LIMIT 9223372036854775807 -- large limit to enable ordering to work
          /* middle select */
      ) AS `t5`
      /* outer select */
  ) AS `posts`
FROM
  `User` AS `t1`
```

### m-n

```prisma
model User {
  id   Int     @id @default(autoincrement())
  name String?

  posts Post[]
}

model Post {
  id    Int     @id @default(autoincrement())
  title String?

  users User[]
}
```

```graphql
{
  findManyUser {
    id
    name
    posts(orderBy: { id: asc }) {
      id
      title
    }
  }
}
```

```sql
SELECT
  `t1`.`id`,
  `t1`.`name`,
  (
    SELECT
      COALESCE(
        JSON_ARRAYAGG(`__prisma_data__`),
        CONVERT('[]', JSON)
      ) AS `__prisma_data__`
    FROM
      (
        SELECT
          JSON_OBJECT('id', `t3`.`id`, 'title', `t3`.`title`) AS `__prisma_data__`
        FROM
          (
            SELECT
              `Post`.*
            FROM
              `_PostToUser` AS `t2`
              INNER JOIN `Post` ON `t2`.`A` = `Post`.`id`
            WHERE
              `t2`.`B` = `t1`.`id`
            ORDER BY
              `Post`.`id` ASC
              /* root */
          ) AS `t3`
        LIMIT
          9223372036854775807
          /* inner */
      ) AS `t4`
      /* outer */
  ) AS `posts`
FROM
  `User` AS `t1`
```

Note: When fetching to-many relations, if any ordering is present, we add a limit of `i64::MAX` which inexplicably forces MySQL to order the aggregated rows.

### CODE REVIEW

File: quaint/src/visitor/mysql.rs [hunk 1]
```diff
@@ -90,6 +90,36 @@ impl<'a> Mysql<'a> {
...
+                (
+                    Some(TypeFamily::Text(_)),
+                    Some("LONGBLOB") | Some("BLOB") | Some("MEDIUMBLOB") | Some("SMALLBLOB") | Some("TINYBLOB")
+                    | Some("VARBINARY") | Some("BINARY") | Some("BIT"),
+                ) => {
+                    self.write("to_base64")?;
+                    self.surround_with("(", ")", |s| s.visit_expression(expr))?;
+
+                    Ok(())
+                }
```
Comment (Weakky) [inline quaint/src/visitor/mysql.rs:107]: We convert binary columns to `base64` to ease coercion in the Query Engine

File: quaint/src/visitor/mysql.rs [hunk 2]
```diff
@@ -90,6 +90,36 @@ impl<'a> Mysql<'a> {
...
+                // Convert floats to string to avoid losing precision
+                (_, Some("FLOAT")) => {
+                    self.write("CONVERT")?;
+                    self.surround_with("(", ")", |s| {
+                        s.visit_expression(expr)?;
+                        s.write(", ")?;
+                        s.write("CHAR")
+                    })?;
+                    Ok(())
+                }
```
Comment (Weakky) [inline quaint/src/visitor/mysql.rs:117]: We convert floats to string to avoid losing precision

File: query-engine/query-structure/src/field/scalar.rs [hunk 3]
```diff
@@ -155,15 +155,22 @@ impl ScalarField {
...
 
-        let nt = connector
-            .parse_native_type(name, args, span, &mut Default::default())
-            .unwrap();
+        let scalar_type = match self.id {
+            ScalarFieldId::InModel(id) => self.dm.walk(id).scalar_type(),
+            ScalarFieldId::InCompositeType(id) => self.dm.walk(id).scalar_type(),
+        };
+
+        let nt = psl_nt.or_else(|| scalar_type.and_then(|st| connector.default_native_type_for_scalar_type(&st)))?;
```
Comment (Weakky) [inline query-engine/query-structure/src/field/scalar.rs:173]: To enable applying special conversion in quaint for specific native types, I updated the `ScalarField::native_type` function to return the default native type in case none are specified. Previously, this function would only return something _if_ a native type was explicitly specified.

File: psl/psl-core/src/builtin_connectors/mongodb/mongodb_types.rs [hunk 4]
```diff
@@ -39,7 +39,7 @@ static DEFAULT_MAPPING: Lazy<HashMap<ScalarType, MongoDbType>> = Lazy::new(|| {
         (ScalarType::Float, MongoDbType::Double),
         (ScalarType::Boolean, MongoDbType::Bool),
         (ScalarType::String, MongoDbType::String),
-        (ScalarType::DateTime, MongoDbType::Timestamp),
+        (ScalarType::DateTime, MongoDbType::Date),
```
Comment (Weakky) [inline psl/psl-core/src/builtin_connectors/mongodb/mongodb_types.rs:42]: The change mentioned above 👆 has rippled in unintended ways. Notably, it surfaced an incorrect default native type for the `DateTime` prisma type. This is why I've changed this.

File: psl/psl-core/src/builtin_connectors/sqlite_datamodel_connector.rs [hunk 5]
```diff
@@ -66,8 +66,8 @@ impl Connector for SqliteDatamodelConnector {
         unreachable!("No native types on Sqlite");
     }
 
-    fn default_native_type_for_scalar_type(&self, _scalar_type: &ScalarType) -> NativeTypeInstance {
-        NativeTypeInstance::new(())
+    fn default_native_type_for_scalar_type(&self, _scalar_type: &ScalarType) -> Option<NativeTypeInstance> {
+        None
     }
```
Comment (Weakky) [inline psl/psl-core/src/builtin_connectors/sqlite_datamodel_connector.rs:71]: Similarly, now that this method is called for all connectors when constructing quaint columns, the function's signature was changed to better signal when a scalar type has no default native type.

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/relation_load_strategy.rs [hunk 6]
```diff
@@ -86,7 +86,9 @@ mod relation_load_strategy {
 
     async fn assert_used_lateral_join(runner: &mut Runner, expected: bool) {
         let logs = runner.get_logs().await;
-        let actual = logs.iter().any(|l| l.contains("LEFT JOIN LATERAL"));
+        let actual = logs
+            .iter()
+            .any(|l| l.contains("LEFT JOIN LATERAL") || (l.contains("JSON_ARRAYAGG") && l.contains("JSON_OBJECT")));
```
Comment (Weakky) [inline query-engine/connector-test-kit-rs/query-engine-tests/tests/new/relation_load_strategy.rs:91]: Since we're not using JOINs to resolve relations on MySQL, this had to be changed with a loose and worse heuristic.

File: query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/mod.rs [hunk 7]
```diff
@@ -378,12 +383,12 @@ pub(crate) fn should_run(
...
         .iter()
-        .filter_map(|c| ConnectorVersion::try_from(*c).ok())
+        .map(|c| ConnectorVersion::try_from(*c).unwrap())
         .collect::<Vec<_>>();
 
     let inclusions = only
         .iter()
-        .filter_map(|c| ConnectorVersion::try_from(*c).ok())
+        .map(|c| ConnectorVersion::try_from(*c).unwrap())
         .collect::<Vec<_>>();
```
Comment (Weakky) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/connector_tag/mod.rs:392]: Unrelated change to make sure that when a version is specified for the `only` / `exclude` attributes in the test suite, if that version is incorrectly specified, it errors instead of silently ignoring it.

File: query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs [hunk 8]
```diff
@@ -89,8 +91,13 @@ fn process_template(template: String, renderer: Box<dyn DatamodelRenderer>) -> S
...
 
-fn render_preview_features(excluded_features: &[&str]) -> String {
-    let excluded_features: Vec<_> = excluded_features.iter().map(|f| format!(r#""{f}""#)).collect();
+fn render_preview_features(excluded_features: &[&str], version: &ConnectorVersion) -> String {
+    let mut excluded_features: Vec<_> = excluded_features.iter().map(|f| format!(r#""{f}""#)).collect();
+
+    // TODO: Remove this once we are able to have version speficic preview features.
+    if version.is_mysql() && !matches!(version, ConnectorVersion::MySql(Some(MySqlVersion::V8))) {
+        excluded_features.push(format!(r#""{}""#, PreviewFeature::RelationJoins));
+    }
```
Comment (Weakky) [inline query-engine/connector-test-kit-rs/query-tests-setup/src/datamodel_rendering/mod.rs:100]: To avoid having to exclude hundreds of tests for MySQL 5.6 & 5.7, we simply do not render the `relationJoins` preview feature for now when rendering the datamodel of each tests. This ensure the `relationLoadStrategy: join` is never used.

File: query-engine/connectors/mongodb-query-connector/src/value.rs [hunk 9]
```diff
@@ -152,7 +152,7 @@ impl IntoBson for (&MongoDbType, PrismaValue) {
 
             // Double
             (MongoDbType::Double, PrismaValue::Int(i)) => Bson::Double(i as f64),
-            (MongoDbType::Double, PrismaValue::Float(f)) => Bson::Double(f.to_f64().convert(expl::MONGO_DOUBLE)?),
+            (MongoDbType::Double, PrismaValue::Float(f)) => bigdecimal_to_bson_double(f)?,
```
Comment (Weakky) [inline query-engine/connectors/mongodb-query-connector/src/value.rs:155]: Side-effect of `ScalarField::native_type` always returning a native type now. This is just a bug-fix that had not been caught before.

File: query-engine/connectors/sql-query-connector/src/database/operations/coerce.rs [hunk 10]
```diff
@@ -133,6 +141,17 @@ pub(crate) fn coerce_json_scalar_to_pv(value: serde_json::Value, sf: &ScalarFiel
...
+                let err =
+                    || build_conversion_error(sf, &format!("Number({n})"), &format!("{:?}", sf.type_identifier()));
+                let i = n.as_i64().ok_or_else(err)?;
+
+                match i {
+                    0 => Ok(PrismaValue::Boolean(false)),
+                    1 => Ok(PrismaValue::Boolean(true)),
+                    _ => Err(err()),
+                }
+            }
```
Comment (Weakky) [inline query-engine/connectors/sql-query-connector/src/database/operations/coerce.rs:154]: MySQL-specific coercion for tinyint (IIRC)

File: query-engine/connectors/sql-query-connector/src/query_builder/select/mod.rs [hunk 11]
```diff
@@ -1,89 +1,142 @@
...
+            Flavour::Postgres | Flavour::Cockroach => {
+                PostgresSelectBuilder::default().build(args, selected_fields, ctx)
+            }
+            _ => unreachable!("Connector does not support joined selects."),
+        }
     }
+}
 
-    pub(crate) fn build(
+pub(crate) trait JoinSelectBuilder {
```
Comment (Weakky) [inline query-engine/connectors/sql-query-connector/src/query_builder/select/mod.rs:38]: To deal with the differences of PG & MySQL (lateral join vs correlated sub-queries), I've went for a trait design with a bunch of unimplemented methods that handle the few differences that we have between both connectors.

It has grown a bit too much to my taste already with the addition of `_count` support, but I can't think of a cleaner way to avoid huge code duplication.

File: query-engine/connectors/sql-query-connector/src/query_builder/select/mod.rs [hunk 12]
```diff
@@ -140,6 +193,14 @@ impl SelectBuilder {
 
             let inner = inner.with_columns(inner_selection.into()).comment("inner select");
 
+            let middle_take = match connector_flavour(&rs.args) {
+                // On MySQL, using LIMIT makes the ordering of the JSON_AGG working. Beware, this is undocumented behavior.
+                // Note: Ideally, this should live in the MySQL select builder, but it's currently the only implementation difference
+                // between MySQL and Postgres, so we keep it here for now to avoid code duplication.
+                Flavour::Mysql if !rs.args.order_by.is_empty() => rs.args.take_abs().or(Some(i64::MAX)),
+                _ => rs.args.take_abs(),
+            };
```
Comment (Weakky) [inline query-engine/connectors/sql-query-connector/src/query_builder/select/mod.rs:202]: 🫠

File: query-engine/connector-test-kit-rs/query-engine-tests/tests/new/relation_load_strategy.rs (hunk 6 above)
Comment (aqrln) [reply query-engine/connector-test-kit-rs/query-engine-tests/tests/new/relation_load_strategy.rs:91]: I think it's fine. Should we maybe now rename the function to something else? Same problem as https://github.com/prisma/prisma-engines/pull/4639#discussion_r1478655075 I guess.

**Review by aqrln — APPROVED** [review]
🎉

File: psl/psl-core/src/builtin_connectors/cockroach_datamodel_connector.rs [hunk 13]
```diff
@@ -60,9 +60,9 @@ const CAPABILITIES: ConnectorCapabilities = enumflags2::make_bitflags!(Connector
     InsertReturning |
     UpdateReturning |
     RowIn |
-    LateralJoin |
     DeleteReturning |
-    SupportsFiltersOnRelationsWithoutJoins
+    SupportsFiltersOnRelationsWithoutJoins |
+    LateralJoin
```
[the remark this answers is not in the document]
Comment (Weakky) [reply psl/psl-core/src/builtin_connectors/cockroach_datamodel_connector.rs:65]: Yes, that's much better actually, because we wouldn't have to match on connector flavours in the top-level dispatch of the SelectBuilder 👍

**Review by laplab — APPROVED** [review]
Had a quick call with Flavian to go over the key parts. LGTM, great work!
